lib/elixir/lib/kernel/typespec.ex

```diff
@@ -283,6 +283,11 @@ defmodule Kernel.Typespec do
       compile_error(caller, message)
     end
 
+    if kind == :opaque and arity == 0 and underspecified?(spec) do
```

Comment:
Maybe we should move the opaque and arity inside the `underspecified?` function?

Reply:
I started with `underspecified?(kind, arity, spec)` but it felt weird that the message have the "opaque" hardcoded. 

But rethinking this part, maybe we can reword the message to include the kind? For example:

```elixir
if underspecified?(kind, arity, spec) do
  message = "@#{kind} type #{name}/#{arity} is underspecified and therefore meaningless"
  :elixir_errors.warn(caller.line, caller.file, message)
end
```

Reply:
LGTM!